parser.go

```diff
@@ -74,24 +74,40 @@ func (p *Parser) ParseWithClaims(tokenString string, claims Claims, keyFunc Keyf
 		}
 	}
 
-	// Lookup key
-	var key interface{}
+	// Decode signature
+	token.Signature, err = p.DecodeSegment(parts[2])
+	if err != nil {
+		return token, newError("could not base64 decode signature", ErrTokenMalformed, err)
+	}
+	text := strings.Join(parts[0:2], ".")
+
+	// Lookup key(s)
 	if keyFunc == nil {
 		// keyFunc was not provided.  short circuiting validation
 		return token, newError("no keyfunc was provided", ErrTokenUnverifiable)
 	}
-	if key, err = keyFunc(token); err != nil {
-		return token, newError("error while executing keyfunc", ErrTokenUnverifiable, err)
-	}
 
-	// Decode signature
-	token.Signature, err = p.DecodeSegment(parts[2])
+	got, err := keyFunc(token)
 	if err != nil {
-		return token, newError("could not base64 decode signature", ErrTokenMalformed, err)
+		return token, newError("error while executing keyfunc", ErrTokenUnverifiable, err)
 	}
 
-	// Perform signature validation
-	if err = token.Method.Verify(strings.Join(parts[0:2], "."), token.Signature, key); err != nil {
+	switch have := got.(type) {
+	case PublicKeyset:
```

Comment:
I suspect folks will try to use `[]jwt.PublicKey` here. What's the rationale for the "public key set" struct instead of asserting for a slice of PublicKey?

I guess it'd enable us to evolve the public key set in a backwards-compatible way if we wanted to attach additional fields.

Reply:
To clarify, you're in favor of using a struct instead of a slice, correct?

Reply:
Yes